### DISCUSSION TITLE
您好，请教一下MVVM中如何让用户确认后再更新数据源？

### DISCUSSION THREAD

**Discussion by BeautyOfCoding** [BYJRK/DotNet-Discussions#60] [MVVM]
我经常会使用WPF做一些参数配置的界面，但是在实际使用过程中我并不想使用TwoWay来绑定，因为一些特殊的场合，比如配置一个报警阈值的参数ThreshValue，后台ViewModel里面有一个线程不断根据ThreshValue来判断某些数值时候超标，如果使用TwoWay绑定那么只要鼠标离开焦点立马就更新数据源，而这个时候用户不小心发现输入错了，但实际上后台ViewModel的ThreshValue已经更新了，我想要的效果是类似于点击一个按钮确认后才能更新数据源，如果使用OneWay却又得不到目标数据，请问这个矛盾如何解决？

**Comment by KiyanYang** [comment]
来自 New Bing:

在绑定机制中，如果想要设置“保存”和“取消”按钮，来保存或取消相应的数据变化，有以下几种可能的方法：

1. 使用双向绑定（TwoWay Binding）和更新触发器（UpdateSourceTrigger），将View中的控件属性与ViewModel中的属性双向绑定，并且设置更新触发器为显式（Explicit），这样就可以在用户点击“保存”按钮时，手动调用BindingExpression.UpdateSource方法来更新ViewModel中的属性值；而在用户点击“取消”按钮时，不做任何操作。
2. 使用单向绑定（OneWay Binding）和命令参数（CommandParameter），将View中的控件属性与ViewModel中的属性单向绑定，并且将控件本身作为命令参数传递给ViewModel中的命令，这样就可以在用户点击“保存”按钮时，获取命令参数中的控件属性值并赋值给ViewModel中的属性；而在用户点击“取消”按钮时，不做任何操作。
3. 使用备份变量（Backup Variable）和撤销更改（Undo Changes），在ViewModel中定义一个备份变量来存储原始的属性值，并且实现一个撤销更改的方法来恢复原始值，这样就可以在用户点击“保存”按钮时，更新备份变量为当前属性值；而在用户点击“取消”按钮时，调用撤销更改的方法来还原属性值

**Reply by BeautyOfCoding** [reply]
万分感谢，我试试！

**Reply by BeautyOfCoding** [reply]
大佬，如果用第二种方法，假设有多个TextBox使用OneWay Binding，但是一个按钮的Command的CommandParameter貌似智能传递一个参数呀，如何把这组的TextBox都传递过去？

**Reply by KiyanYang** [reply]
哈哈，选合适的呗😂我之前用的第三个方法，前两个没用过

**Reply by eijjxx** [reply]
建议用第三个方法，未来你总会遇到有效性检查之类的一些需求，第三个方法比较方便。
第一第二个方法如果按钮是绑定到command的话，会不太mvvm
